json_ref_dict/ref_pointer.py

```diff
@@ -47,8 +47,15 @@ def resolve_remote_with_uri(
         :return: tuple indicating (1) if doc was a ref (the ref URI returned)
             and (2) what that ref value was.
         """
+
+        next_part = None
+        if parts_idx + 1 < len(self.parts):
+            next_part = self.parts[parts_idx + 1]
+
         if not (
-            isinstance(doc, abc.Mapping) and isinstance(doc.get("$ref"), str)
+            isinstance(doc, abc.Mapping)
+            and isinstance(doc.get("$ref"), str)
+            and not next_part in doc
```

Comment:
I think this could produce issues. Consider:

```yaml
foo:
    $ref: #/bar/value
bar:
    $ref: #/baz
    value: "bar_value"
baz:
    value: "baz_value"
```

Then what should `RefDict("file.yaml")["foo"]` resolve to? According to spec, it should resolve to `"baz_value"`, but I think with this change, it would resolve to `"bar_value"`.

Reply:
Yes this is true. I'll have to explore if there's a better way to determine if a reference is pointing within itself.

Reply:
I'm open to suggestions on how to identify that! I have a suspicion it might be an tricky/expensive thing to compute..

Reply:
I've added a test for this scenario (which is currently failing), will look into resolving.

Reply:
I've re-written this with a more complete solution which takes this issue into account.

A `$ref` is now only ignored when the the `RefPointer` being resolved comes across itself, **and** the resolver has more parts of the URI to resolve. This allows the existing `RefPointer` resolve process to continue.

If there is not more to resolve, then the reference is actually circular on itself directly, and the existing behaviour which handles that is triggered (by not ignoring the `$ref`).